### PR TITLE
updated json response to match spec by @davebryson

### PR DESCRIPTION
**Pull request by melekes**
Port of https://github.com/tendermint/go-rpc/pull/19

link to spec: http://www.jsonrpc.org/specification

### CODE REVIEW

File: rpc/lib/server/handlers.go [hunk 1]
```diff
@@ -109,35 +109,35 @@ func makeJSONRPCHandler(funcMap map[string]*RPCFunc, logger log.Logger) http.Han
 		var request types.RPCRequest
 		err := json.Unmarshal(b, &request)
 		if err != nil {
-			WriteRPCResponseHTTPError(w, http.StatusBadRequest, types.NewRPCResponse("", nil, fmt.Sprintf("Error unmarshalling request: %v", err.Error())))
+			WriteRPCResponseHTTP(w, types.RPCParseError(""))
```
Comment (melekes) [inline rpc/lib/server/handlers.go]: I don't like we are loosing error messages. Should we log them instead or expose one more field (like `additional` or something)?

Comment (melekes) [inline rpc/lib/server/handlers.go]: @davebryson what do you think?

Comment (davebryson) [inline rpc/lib/server/handlers.go]: The Error response provides a base message corresponding to the error code.  In my first version of this code, I had included the "data" field in the error that could have been used to add additional info - but I removed it based on earlier feedback.    Logging will probably work just fine.

Comment (melekes) [inline rpc/lib/server/handlers.go]: I have put data field back https://github.com/tendermint/tendermint/pull/521/commits/d75c1e44b89d85cc94fb0fa428f69383632af3b5

**Comment by melekes** [comment]
Looks good now.

File: rpc/lib/client/ws_client.go [hunk 2]
```diff
@@ -111,8 +111,8 @@ func (wsc *WSClient) receiveEventsRoutine() {
 				wsc.ErrorsCh <- err
 				continue
 			}
-			if response.Error != "" {
-				wsc.ErrorsCh <- errors.Errorf(response.Error)
+			if response.Error != nil && response.Error.Message != "" {
```
Comment (odeke-em) [inline rpc/lib/client/ws_client.go]: So what happens to all the other, response.Error's that have empty messages?
Will they all now just fall through and be treated as success?

Perhaps we can tweak this a little e.g
```go
if response.Error != nil {
   switch response.Error.Message {
   case "":
       wsc.ErrorsCh <- errors.Errorf(response.Error)
   default:
       wsc.ErrorsCh <- errors.Errorf(response.Error.Message)
   }
}
```
that way we can catch all errors.
PS: I obviously don't have much context about the code base since I noticed the change from a string to nullable type :)

File: rpc/lib/types/types.go [hunk 3]
```diff
@@ -42,34 +47,69 @@ func ArrayToRequest(id string, method string, params []interface{}) (RPCRequest,
...
+}
 
 type RPCResponse struct {
 	JSONRPC string           `json:"jsonrpc"`
-	ID      string           `json:"id"`
-	Result  *json.RawMessage `json:"result"`
-	Error   string           `json:"error"`
+	ID      string           `json:"id,omitempty"`
+	Result  *json.RawMessage `json:"result,omitempty"`
+	Error   *RpcError        `json:"error,omitempty"`
```
Comment (odeke-em) [inline rpc/lib/types/types.go]: Just a random question: does this change break compatibility for folks consuming types/types?

Comment (melekes) [inline rpc/lib/types/types.go]: yes

File: rpc/lib/client/ws_client.go (hunk 2 above)
Comment (melekes) [inline rpc/lib/client/ws_client.go]: Thanks. Actually, the error message should always be present. I removed the checks.

**Comment by melekes** [comment]
@ebuchman can we merge this?

File: rpc/lib/types/types.go [hunk 4]
```diff
@@ -42,34 +47,69 @@ func ArrayToRequest(id string, method string, params []interface{}) (RPCRequest,
...
+		Error:   &RPCError{Code: code, Message: msg, Data: data},
 	}
 }
 
+func RPCParseError(id string, err error) RPCResponse {
+	return NewRPCErrorResponse(id, -32700, "Parse error. Invalid JSON", err.Error())
+}
+
+func RPCInvalidRequestError(id string, err error) RPCResponse {
+	return NewRPCErrorResponse(id, -32600, "Invalid Request", err.Error())
```
Comment (odeke-em) [inline rpc/lib/types/types.go]: Just a drive-by question, how come the negative status codes?

Comment (davebryson) [inline rpc/lib/types/types.go]: They are from the JSON-RPC spec

Comment (odeke-em) [inline rpc/lib/types/types.go]: Thank you @davebryson, after your comment I then referred to them here http://www.jsonrpc.org/specification#response_object.

**Comment by melekes** [comment]
LGTM

File: rpc/lib/server/handlers.go [hunk 5]
```diff
@@ -228,25 +228,25 @@ func makeHTTPHandler(rpcFunc *RPCFunc, logger log.Logger) func(http.ResponseWrit
 	// Exception for websocket endpoints
 	if rpcFunc.ws {
 		return func(w http.ResponseWriter, r *http.Request) {
-			WriteRPCResponseHTTPError(w, http.StatusMethodNotAllowed, types.NewRPCResponse("", nil, "This RPC method is only for websockets"))
+			WriteRPCResponseHTTP(w, types.RPCInternalError("", errors.New("Trying to use Websocket method in non-ws context")))
```
Comment (ebuchman) [inline rpc/lib/server/handlers.go]: should we be doing JSONRPC error codes for the makeHTTPHandler? yes the response is JSON, but since the request wasn't done via JSONRPC, perhaps we should be returning http error codes ? ie. we only follow the JSONRPC spec when the request is via JSONRPC

File: rpc/lib/types/types.go [hunk 6]
```diff
@@ -42,34 +47,69 @@ func ArrayToRequest(id string, method string, params []interface{}) (RPCRequest,
...
+	Message string `json:"message"`
+	Data    string `json:"data,omitempty"`
+}
 
 type RPCResponse struct {
 	JSONRPC string           `json:"jsonrpc"`
-	ID      string           `json:"id"`
-	Result  *json.RawMessage `json:"result"`
-	Error   string           `json:"error"`
+	ID      string           `json:"id,omitempty"`
```
Comment (ebuchman) [inline rpc/lib/types/types.go]: from the spec:

```
This member is REQUIRED.
It MUST be the same as the value of the id member in the Request Object.
If there was an error in detecting the id in the Request object (e.g. Parse error/Invalid Request), it MUST be Null.
```

I think this means we can't use `omitempty` and that we need to use a pointer for the ID so that it can be encoded properly as `null` rather than empty string ...

File: rpc/lib/server/handlers.go [hunk 7]
```diff
@@ -109,35 +109,35 @@ func makeJSONRPCHandler(funcMap map[string]*RPCFunc, logger log.Logger) http.Han
 		var request types.RPCRequest
 		err := json.Unmarshal(b, &request)
 		if err != nil {
-			WriteRPCResponseHTTPError(w, http.StatusBadRequest, types.NewRPCResponse("", nil, fmt.Sprintf("Error unmarshalling request: %v", err.Error())))
+			WriteRPCResponseHTTP(w, types.RPCParseError("", errors.Wrap(err, "Error unmarshalling request")))
 			return
 		}
```
Comment (ebuchman) [inline rpc/lib/server/handlers.go]: from spec:

```
A Notification is a Request object without an "id" member.
...
 The Server MUST NOT reply to a Notification, including those that are within a batch request.
```

So I think we need to check the request.ID here and not return anything if it's empty

**Comment by ebuchman** [comment]
Oh, I'd also love for @davebryson to get the contributing credits on the first commit here which would otherwise be lost because original PR is in deprecated repo. Can we do some git magic-foo to make that happen ?

**Comment by melekes** [comment]
@ebuchman of course, we can. see  https://github.com/jayphelps/git-blame-someone-else. we can even put Linus Torvalds email if we want to :)

**Comment by adrianbrink** [comment]
And now we know why you should sign your commits. Even though if I remember correctly Linus had some strong opinions about that one too.

**Comment by odeke-em** [comment]
How's it going @melekes? Gentle ping here, as this PR has some merge conflicts and perhaps if we could get it merged in, we'd have our JSON responses matching the spec.

**Comment by melekes** [comment]
I am gonna work on this now.

File: rpc/lib/server/handlers.go (hunk 5 above)
Comment (melekes) [reply rpc/lib/server/handlers.go]: > but since the request wasn't done via JSONRPC

yes, but the format in use is JSONRPC (i.e. both request and response objects contain `jsonrpc` field). I think it is ok to leave the error codes

File: rpc/lib/server/handlers.go (hunk 7 above)
Comment (melekes) [reply rpc/lib/server/handlers.go]: fixed

File: rpc/lib/server/handlers.go [hunk 8]
```diff
@@ -109,35 +109,35 @@ func makeJSONRPCHandler(funcMap map[string]*RPCFunc, logger log.Logger) http.Han
...
 		rpcFunc := funcMap[request.Method]
 		if rpcFunc == nil {
-			WriteRPCResponseHTTPError(w, http.StatusNotFound, types.NewRPCResponse(request.ID, nil, "RPC method unknown: "+request.Method))
+			WriteRPCResponseHTTP(w, types.RPCMethodNotFoundError(request.ID))
 			return
 		}
 		if rpcFunc.ws {
-			WriteRPCResponseHTTPError(w, http.StatusMethodNotAllowed, types.NewRPCResponse(request.ID, nil, "RPC method is only for websockets: "+request.Method))
+			WriteRPCResponseHTTP(w, types.RPCInternalError(request.ID, errors.New("Trying to use Websocket method in non-ws context")))
```
[the remark this answers is not in the document]
Comment (melekes) [reply rpc/lib/server/handlers.go]: yes

File: rpc/lib/types/types.go (hunk 6 above)
Comment (melekes) [reply rpc/lib/types/types.go]: fixed

**Comment by melekes** [comment]
> of course, we can. see https://github.com/jayphelps/git-blame-someone-else. we can even put Linus Torvalds email if we want to :)

this tool does not work properly :(

File: rpc/lib/server/handlers.go (hunk 5 above)
Comment (ebuchman) [reply rpc/lib/server/handlers.go]: Ok. Just to re-iterate, it means it will return 200, and the caller has to check the json for the error. I guess it's fine

**Comment by melekes** [comment]
@ebuchman can we merge it now?
